src/backend/nodes/readfast.c

```diff
@@ -1044,6 +1047,45 @@ _readCollateClause(void)
 	READ_DONE();
 }
 
+static ExtensibleNode *
+_readExtensibleNode(void)
+{
+	const ExtensibleNodeMethods *methods;
+	ExtensibleNode *local_node;
+	const char *extnodename;
+
+	char *str;
+	char **save_strtok = NULL;
+	char **save_begin = NULL;
+
+	READ_STRING_VAR(extnodename);
+	if (!extnodename)
+		elog(ERROR, "extnodename has to be supplied");
+	methods = GetExtensibleNodeMethods(extnodename, false);
+
+	local_node = (ExtensibleNode *) newNode(methods->node_size,
+											T_ExtensibleNode);
+	local_node->extnodename = extnodename;
+
+	READ_STRING_VAR(str);
+
+	/*
+	 * deserialize the private fields
+	 */
+
+	/* set the states for pg_strtok(), let methods->nodeRead() to process str */
+	save_strtok_states(save_strtok, save_begin);
+	set_strtok_states(str, str);
```

Comment:
I am a bit confused here. The tokenisation states seem to be for debugging or am I missing something?
If they are indeed for debug, then I would suggest to include them in ifdef DEBUG blocks to help with readability

Reply:
`pg_strtok_begin` is for debugging as the comments say.

But the static variable `pg_strtok_ptr` is for `pg_strtok()` to memorize the current position while decoding the serialized string.

The trick here is, `methods->nodeRead()` uses the `pg_strtok_ptr` information via `pg_strtok()`, but because the current context is binary context (readfast.c), and the serialized string is not the one `pg_strtok()` was using, we need to provide that `pg_strtok_ptr` information based on the serialized string for `methods->nodeRead()`.

It's time to add more comments for readfuncs/readfast I suppose :)

Reply:
Thanks.